src/payment/strategies/paypal-commerce/paypal-commerce-payment-strategy.ts

```diff
@@ -60,15 +65,40 @@ export default class PaypalCommercePaymentStrategy implements PaymentStrategy {
         return Promise.resolve(this._store.getState());
     }
 
-    private async _getOrderId(methodId: string): Promise<string> {
+    private async _getOrderIdAndShowPopup(methodId: string): Promise<string> {
         const state = this._store.getState();
         const cart = state.cart.getCartOrThrow();
         const { approveUrl, orderId } = await this._paypalCommerceRequestSender.setupPayment(methodId, cart.id);
+        this._approveUrl = approveUrl;
 
-        if (approveUrl) {
-            await this._paypalCommercePaymentProcessor.paymentPayPal(approveUrl);
-        }
+        await this._showPopup();
 
         return orderId;
     }
+
+    private async _showPopup(): Promise<any> {
+        if (this._approveUrl) {
+            await this._paypalCommercePaymentProcessor.paymentPayPal(this._approveUrl);
+        }
+    }
+
+    private async _processSubmitPayment(payment: Payment, methodId: string, attempts = 0): Promise<InternalCheckoutSelectors> {
+        try {
+            return await this._store.dispatch(this._paymentActionCreator.submitPayment({ ...payment }));
+        } catch (error) {
+            const isNotTransactionRejected = !(error instanceof RequestError) || !some(error.body.errors, {code: 'transaction_rejected'});
+
+            if (isNotTransactionRejected || attempts > 2) {
+                return Promise.reject(error);
+            }
+
+            if (!this._approveUrl) {
+                await this._getOrderIdAndShowPopup(methodId);
+            } else {
+                await this._showPopup();
+            }
+
+            return this._processSubmitPayment(payment, methodId, ++attempts);
```

Comment:
3d param could be `attempts + 1`

Reply:
Sure, and could be `++attempts`. It's the same 
Why not `++attempts`?